app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -0,0 +1,185 @@
+require 'command_class'
+
+module Authentication
+  module AuthnAzure
+
+    Log = LogMessages::Authentication::AuthnAzure
+    Err = Errors::Authentication::AuthnAzure
+    # Possible Errors Raised: RoleNotFound, InvalidApplicationIdentity, XmsMiridParseError,
+    # MissingRequiredFieldsInXmsMirid, MissingProviderFieldsInXmsMirid, MissingConstraint,
+    # IllegalConstraintCombinations
+
+    ValidateApplicationIdentity = CommandClass.new(
+      dependencies: {
+        role_class:                 ::Role,
+        resource_class:             ::Resource,
+        validate_azure_annotations: ValidateAzureAnnotations.new,
+        application_identity_class: ApplicationIdentity,
+        logger:                     Rails.logger
+      },
+      inputs:       %i(account service_id username xms_mirid_token_field oid_token_field)
+    ) do
+
+      def call
+        parse_xms_mirid
+        validate_xms_mirid_format
+        token_identity_from_claims
+        validate_azure_annotations_are_permitted
+        extract_application_identity_from_role
+        validate_required_constraints_exist
+        validate_constraint_combinations
+        validate_token_identity_matches_annotations
+      end
+
+      private
+
+      def parse_xms_mirid
+        xms_mirid_hash
+      end
+
+      def xms_mirid_hash
+        @xms_mirid_hash ||= parsed_xms_mirid
+      end
+
+      # we expect the xms_mirid claim to be in the format of /subscriptions/<subscription-id>/...
+      # therefore, we ignore the first slash of the xms_mirid claim and group the entries in key-value pairs
+      # according to fields we need to retrieve from the claim.
+      # ultimately, transforming "/key1/value1/key2/value2" to {"key1" => "value1", "key2" => "value2"}
+      def parsed_xms_mirid
+        split_xms_mirid = @xms_mirid_token_field.split('/')
+
+        if split_xms_mirid.first == ''
+          split_xms_mirid = split_xms_mirid.drop(1)
+        end
+
+        index = 0
+        split_xms_mirid.each_with_object({}) do |property_name, xms_mirid_hash|
+          property_value_index = index + 1
+
+          case property_name
+          when "subscriptions"
+            xms_mirid_hash["subscriptions"] = split_xms_mirid[property_value_index]
+          when "resourcegroups"
+            xms_mirid_hash["resourcegroups"] = split_xms_mirid[property_value_index]
+          when "providers"
+            xms_mirid_hash["providers"] = split_xms_mirid[property_value_index, 3]
+          end
+          index += 1
+        end
```

Comment:
This section here could be made much more idiomatic.  if i'm understanding the code the `split_xms_mirid` is an array that looks something like this:
```ruby
arr = ['subscriptions', 'foo', 'resourcegroups', 'bar', 'providers', 'one', 'two', 'three']
```
with the hash keys and their values in adjacent elements, with providers containing 3 elements after it.    if so, here's a more functional alternative to the above logic:

```ruby
# Slice points:
#        v                      v                        v
# arr = ["subscriptions", "foo", "resourcegroups", "bar", "providers", "one", "two", "three"]

arr = ["subscriptions", "foo", "resourcegroups", "bar", "providers", "one", "two", "three"]

xms_hash = arr
  .slice_before(/subscriptions|resourcegroups|providers/)
  .each_with_object({}) do |part, hash|
    key, val = part.first, part.drop(1)
    val = val.first if val.length == 1 
    hash[key] = val
end

p xms_hash
#=> {"subscriptions"=>"foo", "resourcegroups"=>"bar", "providers"=>["one", "two", "three"]}
```
Additionally, since there's enough logic here and with the validations you do, I'd strongly consider making `XmsMirid` into  its own value object.  You could keep it in this same file if you want, but it would take a string in its constructor, do all the parsing and validation internally, and expose methods `subscriptions`, `resourcegroups`, and `providers`.

Reply:
Beautiful!